File: Packages/StackEngineCore/Tests/StackEngineCoreTests/OutputTransformTests.swift
import XCTest
import simd
@testable import StackEngineCore

final class OutputTransformTests: XCTestCase {
    func testSRGBRoundTripWithinQuantization() {
        // linear → sRGB8 → linear should return ~the original (8-bit quantization tolerance).
        let img = PixelImage(width: 3, height: 1, pixels: [
            SIMD3<Float>(0.0, 0.25, 0.5),
            SIMD3<Float>(0.5, 0.75, 1.0),
            SIMD3<Float>(0.1, 0.2, 0.9),
        ])
        let back = OutputTransform.decodeSRGB8(OutputTransform.encodeSRGB8(img), width: 3, height: 1)
        XCTAssertEqual(back.pixels.count, 3)
        for i in 0..<3 {
            for ch in 0..<3 {
                XCTAssertEqual(back.pixels[i][ch], img.pixels[i][ch], accuracy: 0.006, "pixel \(i) ch \(ch)")
            }
        }
    }

    func testSRGBEncodingKnownValues() {
        let img = PixelImage(width: 3, height: 1, pixels: [
            SIMD3<Float>(0, 0, 0),
            SIMD3<Float>(1, 1, 1),
            SIMD3<Float>(0.5, 0.5, 0.5),
        ])
        let bytes = OutputTransform.encodeSRGB8(img) // RGBA, 4 bytes/pixel
        XCTAssertEqual(bytes.count, 3 * 4)
        // black -> 0, alpha 255
        XCTAssertEqual(bytes[0], 0); XCTAssertEqual(bytes[3], 255)
        // white -> 255
        XCTAssertEqual(bytes[4], 255)
        // linear 0.5 -> sRGB ~0.7353 -> ~188
        XCTAssertEqual(Int(bytes[8]), 188, accuracy: 1)
    }
}

private func XCTAssertEqual(_ a: Int, _ b: Int, accuracy: Int,
                            file: StaticString = #filePath, line: UInt = #line) {
    XCTAssertLessThanOrEqual(abs(a - b), accuracy, file: file, line: line)
}

File: android/stackengine/src/test/kotlin/com/jdpneto/stackengine/OutputTransformTests.kt
package com.jdpneto.stackengine

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class OutputTransformTests {

    @Test
    fun testSRGBRoundTripWithinQuantization() {
        // linear → sRGB8 → linear should return ~the original (8-bit quantization tolerance).
        val img = PixelImage(3, 1, FloatArray(9).also { a ->
            // pixel 0
            a[0] = 0.0f; a[1] = 0.25f; a[2] = 0.5f
            // pixel 1
            a[3] = 0.5f; a[4] = 0.75f; a[5] = 1.0f
            // pixel 2
            a[6] = 0.1f; a[7] = 0.2f;  a[8] = 0.9f
        })
        val back = OutputTransform.decodeSRGB8(OutputTransform.encodeSRGB8(img), 3, 1)
        assertEquals(3, back.pixelCount)
        for (i in 0 until 3) {
            for (ch in 0 until 3) {
                assertEquals(
                    img.pixels[i * 3 + ch],
                    back.pixels[i * 3 + ch],
                    absoluteTolerance = 0.006f,
                    message = "pixel $i ch $ch"
                )
            }
        }
    }

    @Test
    fun testSRGBEncodingKnownValues() {
        val img = PixelImage(3, 1, FloatArray(9).also { a ->
            // pixel 0: black
            a[0] = 0f; a[1] = 0f; a[2] = 0f
            // pixel 1: white
            a[3] = 1f; a[4] = 1f; a[5] = 1f
            // pixel 2: linear 0.5
            a[6] = 0.5f; a[7] = 0.5f; a[8] = 0.5f
        })
        val bytes = OutputTransform.encodeSRGB8(img)  // RGBA, 4 bytes/pixel
        assertEquals(3 * 4, bytes.size)
        // black -> 0, alpha 255
        assertEquals(0, bytes[0].toInt() and 0xFF)
        assertEquals(255, bytes[3].toInt() and 0xFF)
        // white -> 255
        assertEquals(255, bytes[4].toInt() and 0xFF)
        // linear 0.5 -> sRGB ~0.7353 -> ~188
        val val8 = bytes[8].toInt() and 0xFF
        assertTrue(kotlin.math.abs(val8 - 188) <= 1, "expected ~188, got $val8")
    }

    @Test
    fun testEncodeIsNaNAndInfSafe() {
        // A non-finite pixel (e.g. from degenerate upstream math) must not trap; it
        // encodes to 0 (NaN) / 255 (+Inf) / 0 (-Inf) instead of crashing the export.
        val img = PixelImage(3, 1, FloatArray(9).also { a ->
            a[0] = Float.NaN;              a[1] = Float.NaN;              a[2] = Float.NaN
            a[3] = Float.POSITIVE_INFINITY; a[4] = Float.POSITIVE_INFINITY; a[5] = Float.POSITIVE_INFINITY
            a[6] = Float.NEGATIVE_INFINITY; a[7] = Float.NEGATIVE_INFINITY; a[8] = Float.NEGATIVE_INFINITY
        })
        val bytes = OutputTransform.encodeSRGB8(img)
        assertEquals(0,   bytes[0].toInt() and 0xFF)   // NaN → 0
        assertEquals(255, bytes[4].toInt() and 0xFF)   // +Inf → clamped to 1 → 255
        assertEquals(0,   bytes[8].toInt() and 0xFF)   // -Inf → clamped to 0 → 0
    }
}
